File: Gramatica3.py
def Gramatica3(cadena):
    # Verificar que solo tenga 'a' y 'b'
    if not all(c in "ab" for c in cadena):
        return False
    
    num_a = cadena.count('a')
    num_b = cadena.count('b')
    
    # Todos los 'a' antes que 'b'
    if cadena != 'a' * num_a + 'b' * num_b:
        return False
    
    # Condición: num_b = num_a + 1 y al menos 1 'a'
    return num_a > 0 and num_b == num_a + 1


with open("G3.txt","r") as archivo:
    for linea in archivo:
        cadena=linea.strip()
        if Gramatica3(cadena):
            print(f"{cadena}: acepta")
        else:
            print(f"{cadena}: No acepta")   

File: Gramatica3.c
#include <stdio.h>
#include <string.h>

int Gramatica3(char *cadena) {
    int num_a = 0, num_b = 0;
    int len = strlen(cadena);
    int fase_b = 0; // 0 = aun en 'a', 1 = en 'b'
    
    for (int i = 0; i < len; i++) {
        if (cadena[i] == 'a') {
            if (fase_b) return 0; // 'a' después de 'b'
            num_a++;
        } else if (cadena[i] == 'b') {
            fase_b = 1;
            num_b++;
        } else {
            return 0; // Caracter inválido
        }
    }

    // Condición: n > 0 y num_b = num_a + 1
    return (num_a > 0) && (num_b == num_a + 1);
}

int main() {
    char cadena[100];
    printf("Ingresa la cadena: ");
    scanf("%s", cadena);

    if (Gramatica3(cadena))
        printf("acepta\n");
    else
        printf("NO acepta\n");

    return 0;
}
